### PR TITLE
feat: Fix impersonated service account parsing exception

### PR DESCRIPTION
**Pull request by blue-hope**
### Discussion

  * related to #1861 

### Testing

  * test for checking impersonated service account added.

### API Changes

  * no change on public API

RELEASE NOTE: Added support for initializing the SDK with service account impersonation in Application Default Credentials.

### CODE REVIEW

**Comment by blue-hope** [comment]
Can I get any feedback on this? Or discussion on #1861 ?

**Comment by nathan-stable-auto** [comment]
@blue-hope thanks for the PR!

@lahirumaramba I notice you self-assigned the related issue https://github.com/firebase/firebase-admin-node/issues/1703 - would you be able provide any input on how to move this forward? Thanks very much!

**Review by branks — APPROVED** [review]

**Comment by blue-hope** [comment]
@lahirumaramba I hope this feature will be dealt with soon and I would be so appreciated if you could give us a feedback or review.

**Comment by lahirumaramba** [comment]
Thank you everyone for your patience! Hey @blue-hope Thank you so much for the contribution! Just adding this note to let you know that I have started the review process. We also plan to migrate the credentials implementation to `google-auth-library-nodejs` (see https://github.com/firebase/firebase-admin-node/issues/1377) and once that is done firebase-admin will reach feature parity with `google-auth-library-nodejs`. In the meantime, I think it is reasonable to merge this change.

**Comment by IchordeDionysos** [comment]
@lahirumaramba just a side-note on the migration to `google-auth-library-nodejs`:
Right now, it's possible to do the following using a local service account:

Run: `gcloud auth application-default login` 

```
initializeFirebase({project: 'project'});

getFirestore().doc('foo/bar').get();
```

However, trying to authenticate with Identity Toolkit / Firebase Auth would fail:

```
initializeFirebase({project: 'project'});

getAuth().getUser('uid'); // -> throws an error
```

Would the migration to `google-auth-library-nodejs` library fix this problem?

File: test/unit/app/credential-internal.spec.ts [hunk 1]
```diff
@@ -424,6 +424,13 @@ describe('Credential', () => {
       expect(c).to.be.an.instanceof(ServiceAccountCredential);
     });
 
+    it('should return a ImpersonatedCredential with impersonated GOOGLE_APPLICATION_CREDENTIALS set', () => {
+      process.env.GOOGLE_APPLICATION_CREDENTIALS
+        = path.resolve(__dirname, '../../resources/mock.impersonated_key.json');
+      const c = getApplicationDefault();
+      expect(c).to.be.an.instanceof(ImpersonatedServiceAccountCredential);
+    });
+
```
Comment (lahirumaramba) [inline test/unit/app/credential-internal.spec.ts:433]: Let's add another test to `httpAgent`:

```
  it('ImpersonatedServiceAccountCredential should use the provided HTTP Agent', () => {
    const agent = new Agent();
    const c = new ImpersonatedServiceAccountCredential(MOCK_IMPERSONATED_TOKEN_CONFIG, agent);
    return c.getAccessToken().then((token) => {
      expect(token.access_token).to.equal(expectedToken);
      expect(stub).to.have.been.calledOnce;
      expect(stub.args[0][0].httpAgent).to.equal(agent);
    });
  });
```

Comment (lahirumaramba) [reply test/unit/app/credential-internal.spec.ts:433]: Actually, I think we can improve the test coverage for `ImpersonatedServiceAccountCredential`. Please see `describe('RefreshTokenCredential', () => {` for an example.

Let's also add another test in `describe('isApplicationDefault()', () => {`

```
  it('should return true for ImpersonatedServiceAccountCredential loaded from GOOGLE_APPLICATION_CREDENTIALS', () => {
    process.env.GOOGLE_APPLICATION_CREDENTIALS = path.resolve(__dirname, '../../resources/mock.impersonated_key.json');
    const c = getApplicationDefault();
    expect(c).is.instanceOf(ImpersonatedServiceAccountCredential);
    expect(isApplicationDefault(c)).to.be.true;
  });

 it('should return false for explicitly loaded ImpersonatedServiceAccountCredential', () => {
    const c = new ImpersonatedServiceAccountCredential(mockImpersonatedAccount);
    expect(isApplicationDefault(c)).to.be.false;
  });
```

**Review by lahirumaramba — CHANGES_REQUESTED** [review]
Hi @blue-hope, Thank you again for your contributions. It looks pretty good overall... I think we can improve the test coverage. Added a few comments and one question. Please take a look. Thanks!

**Comment by lahirumaramba** [comment]
> Thank you so much for reviewing. (+ Just curious, when will that `google-auth-library-nodejs` implementation be finished?)

I can't promise a timeline for this, but we are starting the initial work this year. It will be a but complex migration so we plan to perform it in multiple stages. We will use #1377 to track any progress.

**Comment by lahirumaramba** [comment]
> @lahirumaramba just a side-note on the migration to `google-auth-library-nodejs`: Right now, it's possible to do the following using a local service account:
> 
> Run: `gcloud auth application-default login`
> 
> ```
> initializeFirebase({project: 'project'});
> 
> getFirestore().doc('foo/bar').get();
> ```
> 
> However, trying to authenticate with Identity Toolkit / Firebase Auth would fail:
> 
> ```
> initializeFirebase({project: 'project'});
> 
> getAuth().getUser('uid'); // -> throws an error
> ```
> 
> Would the migration to `google-auth-library-nodejs` library fix this problem?

Hi @IchordeDionysos, This looks a bit strange. `gcloud auth application-default login` should set `GCLOUD_CREDENTIAL_PATH` and load ADC. If this is reproducible, please file a new issue with a code sample and any error logs. Thanks!

File: src/app/credential-internal.ts [hunk 2]
```diff
@@ -474,9 +573,10 @@ function getDetailFromResponse(response: HttpResponse): string {
   return response.text || 'Missing error payload';
 }
 
-function credentialFromFile(filePath: string, httpAgent?: Agent): Credential {
-  const credentialsFile = readCredentialFile(filePath);
+function credentialFromFile(filePath: string, httpAgent?: Agent, ignoreMissing?: boolean): Credential | null {
+  const credentialsFile = readCredentialFile(filePath, ignoreMissing);
```
[the remark this answers is not in the document]
Comment (blue-hope) [reply src/app/credential-internal.ts:577]: Sure, so I passed `ignoreMissing` as false when check `GOOGLE_APPLICATION_CREDENTIALS` on line 486. So we still can encounter `FirebaseAppError` when an incorrect file path is passed.

File: test/resources/mock.impersonated_key.json [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+    "delegates": [],
+    "service_account_impersonation_url": "",
+    "source_credentials": {
+      "client_id": "client_id",
+      "client_secret": "client_secret",
+      "refresh_token": "refresh_token",
+      "type": "authorized_user"
+    },
+    "type": "impersonated_service_account"
+  }
```
[the remark this answers is not in the document]
Comment (blue-hope) [reply test/resources/mock.impersonated_key.json:11]: Thanks!

**Review by lahirumaramba — APPROVED** [review]
Thanks for addressing the review feedback! LGTM!
Let me know if you can add the missing unit tests. As an alternative, we can merge this as it is and I am happy to add the missing tests in a follow-up PR.

**Comment by blue-hope** [comment]
@lahirumaramba I'll add missing tests and notice to you before merge, thanks!

**Comment by blue-hope** [comment]
@lahirumaramba New test added. I missed checking implicit discovered credential, Thanks a lot.

**Review by lahirumaramba — APPROVED** [review]
Thank you @blue-hope! LGTM!

**Comment by blue-hope** [comment]
@lahirumaramba When this will be released? 😄

**Comment by Klaitos** [comment]
Definitively need this

**Comment by lahirumaramba** [comment]
Thank you @blue-hope for your contribution. This feature is now included in the [v11.5.0 release](https://firebase.google.com/support/release-notes/admin/node#11.5.0).

Thanks folks for your patience on this! Try out the new feature if you get a chance and let us know what you think.
If you encounter any issues, please open a new issue on Github. Thank you!

**Pull request merged**